crates/matrix-sdk/src/widget/README.md

```diff
@@ -0,0 +1,235 @@
+    action: $ACTION,
+    widgetId: $WIDGET_ID,
+    data: {},
+    response: { ... }
+}
+```
+
+The `api` field is required on all requests and must be either `"fromWidget"` or `"toWidget"` depending upon the direction
+of messaging.
+
+The "requestId" field should be unique and included in all requests. This field has been omitted from all of the
+following examples, for the sake of brevity.
+
+The "action" determines the format of the request and response. All actions can return an error response.
+
+The "widgetId" field denotes the widget that the request originates from and is required.
+
+Additional data can be sent as arbitrary fields. However, typically the "data" object should be used.
+
+A success response is an object with zero or more keys where none of keys are “error”.
+
+An error response is a "response" object which consists of a sole "error" key to indicate an error.
+
+They look like:
+
+```json
+{
+    ...
+    error: {
+        message: "Unable to invite user into room.",
+        <Original Error Object>
+    }
+}
+```
+
+The "message" key should be a human-friendly string.
+
+## `WidgetDriver` internals
+
+Internally the driver consists of two main message types:
+
+- [`machine::IncomingMessage`] collects all possible "input" information which can come from the [`MatrixDriver`]
+  or the [`WidgetDriverHandle`] channel, that passes over the actions from the widget.
+  - `WidgetMessage`: An incoming raw message from the widget.
+  - `MatrixDriverResponse`: A response to a request from the `WidgetMachine` to the `MatrixDriver`.
+    For example if the `WidgetMachine` requests the `MatrixDriver` to get approved capabilities,
+    the response would contain what capabilities actually were approved.
+  - `MatrixEventReceived`: The `MatrixDriver` notified the `WidgetMachine` of a new matrix event.
+- [`machine::Action`] describes an operation that the `WidgetDriver` wants to perform:
+  - `SendToWidget`: Send a raw message to the widget.
+  - `MatrixDriverRequest`: A command sent from the client widget API state machine to the
+    client (driver). Once the command is executed,
+    the client will typically generate an `Event` with the result.
+  - `Subscribe`: Subscribe to the events that the widget capabilities allow,
+    in the _current_ room, i.e. a room which this widget is instantiated with.
+    The client is aware of the room.
+  - `Unsubscribe`: Unsubscribe from the events that the widget capabilities allow,
+    in the _current_ room. Symmetrical to `Subscribe`.
+
+```ascii
+                                       Public API ──╮                                                      
+                                                    ▼──────────────────────────────────────────────────┐   
+                                                    │ WidgetDriver                                     │   
+┌────────────────┐  ToWidget response: -            │            ┌──────────────────────────────────┐  │   
```

Comment:
Nooo, everything is centered in the middle except `WidgetMachine` :smiling_face_with_tear: